(Almost) Full Exposure of Voicemeeter Remote API

pblivingston opened the pull request:
## What
Expand and refine the Voicemeeter PowerShell API surface: more parameters/commands on strips and buses, new feature classes (Option, Preset, Fx, Patch), EQ enhancements, VAIO extension support, and Potato Pitch controls. README and CHANGELOG updated.

## Why
- Broader control and parity with Voicemeeter features across variants
- Clearer, less redundant API surface
- Groundwork for future VMRAPI additions (e.g., presets)

## Key additions and changes

- Bus
  - Bus[i].sel (select strip gainlayers with Strip[i].gain)
  - Bus[i].monitor (Monitor On Select)
  - Bus[i].vaio on physical buses (VAIO extension)

- Strip
  - Strip[i].audibility (Voicemeeter Basic)
  - Strip[i].vaio on physical strips (VAIO extension)
  - Strip[i].denoiser.threshold
  - EQ gains on virtual strips: bass/low, mid/med, treble/high
  - Karaoke alias: Strip[i].karaoke for Strip[i].k
  - Pitch (Potato) via new Pitch class:
    - Strip[i].Pitch.on, .drywet, .pitchvalue, .loformant, .medformant, .hiformant

- Commands, Buttons, and Recorder
  - Command.save($filepath), Command.reset
  - Button[i].color (write-only, via Command.Button[i].color)
  - Recorder.eject (via Command.eject)
  - README now mentions Recorder.replay and Recorder.gain

- New classes and API structure
  - Preset class (foundation for store/overwrite, name/comment, load/save); Preset[i].recall implemented
  - Fx class: Reverb.on/ab; Delay.on/ab
  - Patch class: asio[i], OutA2[i]–OutA5[i], composite[i], insert[i], postFaderComposite, postFxInsert
  - Option class:
    - sr, asiosr, delay[i]
    - Buffer: mme, wdm, ks, asio
    - Mode: exclusif, swift
    - monitorOnSel, sliderMode
  - New IRemote and IndexedIRemote base classes to consolidate remote/setter/getter/etc.
  - Alternate gainlayer pattern supported: Strip.gainlayer[i] (existing pattern still works)

- EQ enhancements (Bus.Eq and Strip.Eq)
  - Eq.save($filename), Eq.load($filename)
  - Eq.Channel[i].Cell[j].{on,type,f,gain,q}

### Changes
- Strip[i].limit is now float (GUI truncates visually, value is not actually integer)
- Strip[i].mono is alias for Strip[i].mc on virtual strips (aligns with Python module)
- GainLayers now based on KindMap (only added for Potato)
- Strip and Bus now use IndexedIRemote base class

### Removed
- Bus[i].limit from README (does not exist)

### Fixed
- MacroButton index range in README
- Parenthesis in Strip.AppMute string
- Vban.stream.sr variable corrected (`$this._port` → `$this._sr`)

## Notes / compatibility
- Mostly additive; one potential surface change: Strip[i].limit type changed to float.
- CHANGELOG ([Unreleased]) and README are updated.

## Remaining before Ready for review
- [x] Update/expand tests for new features
- [x] Run tests and investigate/resolve failures
  - [x] All boolean tests pass
  - [x] All integer tests pass
  - [x] All float tests pass
  - [x] All string tests pass
  - [x] All fade tests pass
  - [x] All eq tests pass
  - [x] All special tests pass
- [x] Cleanup/streamline to reduce redundancy
- [x] Re-run tests and finalize docs

## Questions for reviewers
- Are the new parameter/command names and structure acceptable?
- New/adjusted aliases (e.g., mono → mc on virtual, karaoke for k, bass/low, mid/med, treble/high): okay? Any you’d like removed or added?
- Changing Strip[i].limit to float: okay? Is a migration note needed?
- Preset/Fx/Patch/Option/IRemote/IndexedIRemote: class names and placement okay?

pblivingston committed:
bus select exposed

exposes bus[i].sel for selecting the strip gainlayers to be changed with strip[i].gain
alternative to explicitly using strip[i].gainlayer[j] for each gainlayer

pblivingston committed:
monitor, vaio

exposed bus.monitor and bus.vaio
updated readme, removed bus.limit as it does not exist.

pblivingston committed:
Update bus.ps1

correction: 'vaio' is only available on physical buses

pblivingston committed:
more strip parameters

- changed strip.limit to float (the value is visually truncated in the GUI)
- strip.mono is exclusive to physical strips
- strip.audibility added to physical strips for Voicemeeter Standard
- strip.vaio added to physical strips for VAIO Extension
- strip.denoiser.threshold added
- strip.pitch parameters added to physical strips for Voicemeeter Potato
- strip.eqgain1-3 added to virtual strips

pblivingston committed:
reset, save, readme updates

- command.save($filepath) added
- command.reset added
- corrected macrobutton index range in readme
- added recorder.replay and recorder.gain to readme

pblivingston committed:
macrobutton color

added write-only button.color

pblivingston committed:
presets

- added preset objects with 'recall' action
- futureproofing in case vmr api adds: store/overwrite, name, comment, load(filename), save(filename)

pblivingston committed:
Update CHANGELOG.md

pblivingston committed:
recorder.eject

uses Command.eject

pblivingston committed:
corrections to match python module

- moved 'mono' back to [Strip] (silly me, not thinking about back compat)
- changed eqgain exposure to match python module
- added additional 'low' alias for eqgain1

pblivingston committed:
virtual strip aliases

- properties -> aliases to match python module
- override inherited 'mono', matching python module
- added 'karaoke' alias

pblivingston committed:
Fx params

- renamed presets.ps1 to misc.ps1
- added Fx class with
  - Reverb.on
  - Reverb.ab
  - Delay.on
  - Delay.ab

pblivingston committed:
fix fx

- correct class
- remove from basic and banana

pblivingston committed:
GainLayers

- GainLayers are now based on KindMap so they're only added for Potato
- this is also more flexible for possible future changes to Voicemeeter

pblivingston committed:
Patch commands

- New Patch class
  - Patch.asio[i]
  - Patch.OutA2[i]-OutA5[i]
  - Patch.composite[i]
  - Patch.insert[k]
  - Patch.postFaderComposite
  - Patch.postFxInsert
- update README & CHANGELOG

pblivingston committed:
fix patch

had left patch command add...member calls in the for loops

pblivingston committed:
gainlayer aliases

- parenthesis in AppMute string
- moved virtual strip aliases to common hidden block
- changed gainlayers to have brackets around the bus index 'gainlayer[i]' for consistency with other commands/params and so AddFloatMembers can be used directly
- added old gainlayer pattern (gainlayer0..gainlayer7) as an alias for back compat

pblivingston committed:
fix Vban.stream.sr

_port -> _sr

pblivingston committed:
IRemote, Option

- IRemote & IndexedIRemote classes
- misc.ps1 classes now inherit IRemote classes
- Option class, commands

pblivingston committed:
eq commands

- Strip and Bus now use IndexedIRemote base class
- EQ commands added, # of channels based on KindMap
- Eq.Load and Eq.Save will now load/save eq using Command.LoadBusEq[i], etc.

pblivingston committed:
Eq done, bool wip

- eq done
- strip bools done
- bus bools wip

pblivingston committed:
fix bus.mono

change bus.mono to int as it takes 0, 1, 2

pblivingston committed:
Merge branch 'more-params-cmds' into test-updates

pblivingston committed:
strip & bus bools

- strip booleans done
- bus booleans done

pblivingston committed:
bool & command

- general boolean tests done
- special command tests done

pblivingston committed:
vban wip

pblivingston committed:
vban complete

pblivingston committed:
int wip

pblivingston committed:
fix bpsidechain

- bpsidechain float -> int

pblivingston committed:
Merge branch 'more-params-cmds' into test-updates

pblivingston committed:
int wip 2

- strip done
- bus done
- patch done

pblivingston committed:
fix vban route range

0..8 -> 0..7

pblivingston committed:
Merge branch 'more-params-cmds' into test-updates

pblivingston committed:
ints done

option ints

pblivingston committed:
fix delays

int -> custom handling to return second decimal place

pblivingston committed:
Merge branch 'more-params-cmds' into test-updates

pblivingston committed:
floats wip

strip floats will be done after potato audibility/pitch

onyx-and-iris commented:
Hi, thanks for taking the time and interest in this module! I will take a good look over this in the coming days.

pblivingston commented:
I'm fairly new to PowerShell, so critiques, corrections, suggestions, etc. are very welcome!

pblivingston committed:
float, string, fade done

todo: recorder

pblivingston committed:
recorder done; all done

grouped command, vban, and some recorder in 'special tests'

pblivingston committed:
module path

change path so we can run from /tests/

pblivingston committed:
fix syntax

pblivingston committed:
mono conflict

move mono scriptproperty to physical strip

pblivingston committed:
fix tests

ungroup tests because i'm dumb and it crashed my PC

pblivingston committed:
Update higher.Tests.ps1

pblivingston committed:
ParamArray

things like Patch.Insert[i] didn't work the way i had implemented them before - using new ParamArray classes should provide a robust way to interact with these properties
- strip[i].gainlayer[j]
- option.delay[i]
- patch.asio[i]
- patch.outA2-A5[i]
- patch.composite[i]
- patch.insert[i]

pblivingston committed:
Revert "ParamArray"

This reverts commit 1b1e2f7a74afec2f438bbc5deb99382566a740e3.

pblivingston committed:
fix asio crash when testing

- commented Option.asiosr and Option.buffer.asio because an active ASIO driver may crash the PC
- these commands DO work (tested without an ASIO device selected), but exercise caution when testing/using them

took me too many crashes to find out what was going on and at one point my windows PIN data got corrupted - needed to delete the folder in recovery mode to even be able to set a new one.
again, exercise caution.

onyx-and-iris commented on lib/meta.ps1:
I would prefer meta to include only code that generates other code, so if we can move the IRemote classes out of here, perhaps into iremote.ps1.

We should be able to overload the constructor here I believe? Removing the need for an IndexedIRemote class. For example:

```powershell
class IRemote {
    [int]$index
    [Object]$remote

    IRemote ([Object]$remote) {
        $this.remote = $remote
    }

    IRemote ([int]$index, [Object]$remote) {
        $this.index = $index
        $this.remote = $remote
    }
    ...
}
```

pblivingston replied:
[0589ddc](https://github.com/onyx-and-iris/voicemeeter-api-powershell/pull/5/commits/0589ddca3dad499ce6c9eb5d0b61891cd2e92e67) consolidates and moves IRemote to iremote.ps1 and adds arraymembers.ps1

pblivingston committed:
IRemote, ArrayMembers

- moved IRemote to iremote.ps1 and added constructor overload
- fixing array-like params with ArrayMember classes modeled after Recorder.ArmBus[i]|ArmStrip[i]
- still need to do patch.outA2[i], etc. and strip[i].gainlayer[j] and include legacy support for gainlayer0, etc.

pblivingston committed:
Update readme, changelog, and tests

update for ArrayMembers

pblivingston committed:
comment Button[i].color

it 'works' but only takes 0 or 1, not the full 0..8
can leave commented in case vmr api gets updated

pblivingston committed:
gainlayers

- restore AddGainlayerMembers for gainlayer0, etc.
- add gainlayer FloatArrayMembers

pblivingston committed:
test adjustments

- all bool tests pass
- all int tests pass
- all float tests pass
- all string tests pass
- all fade tests pass
- all eq tests pass

TODO:
- special tests

onyx-and-iris commented on CHANGELOG.md:
Yes I've raised this issue already with Vincent (a few times actually) so they are aware of the problem. It's the reason I didn't merge https://github.com/onyx-and-iris/voicemeeter-api-python/commit/46e7ffe4783257bed6c19cbcdca52bb5fa5f6967 into the python interface.

pblivingston committed:
all special tests pass

- command.showvbanchat and .lock are write-only
- fixed accepted recorder.channel values
- vban.enable is not write-only
- instream[i].quality and .route are not read-only
- instream/outstream.port exists on all 16 streams, but vban.instream[0].port is used for all streams
- vban.port now sets and gets vban.instream[0].port for clarity
this could be migrated to set all ports and get the default 6980 if per-stream ports are supported in the future
- leaving 'port' on streams for legacy support & in case per-stream ports are supported in the future
- removed tests with no way to verify; will test these manually

pblivingston commented:
Recorder has both Recorder.mode.loop and Recorder.loop, which just calls Recorder.mode.loop. Recorder.loop is not documented in the README.
- Does Recorder.loop exist for legacy behavior?
- Should it be documented/removed?

onyx-and-iris commented:
Yes it was added before RecordMode class was added, it was kept to avoid introducing a breaking change. If we remove it a major version bump will be in order. We might be able to do that though, since I've already been considering removing the old gainlayer properties.

I'm probably going to ask you to break up this PR before I can consider merging it btw, because it's touching too many files. As well as that it would be useful to show changes added along with the relevant tests incrementally. We can discuss that when you've finished making all the changes to this PR branch.

pblivingston commented:
Okay, sounds good. I'll leave Recorder.loop in for now until we start finalizing things.

pblivingston committed:
asio out members

pblivingston committed:
command, recorder, vban

command:
- IRemote base class
Recorder:
- IRemote base class
- armstrip/armbus -> ArrayMember
Vban:
- IRemote base class
- renamed Vban -> VbanStream
- removed redundant 'VbanInstream'/'VbanOutstream'

pblivingston committed:
fix eject

pblivingston committed:
reduce ForEach-Object piping

removed piping integers to ForEach-Object to reduce overhead

if ForEach-Object is preferred for readability, can revert and change other similar occurrences for consistent style

pblivingston committed:
Create eq.ps1

eq classes

pblivingston committed:
bus eq, strip eq

pblivingston committed:
Revert "reduce ForEach-Object piping"

This reverts commit e51ccf3166c302801791f8d0b9ee7d6dc9dc2dc3.

pblivingston committed:
Merge branch 'more-params-cmds' into consolidate-eq

pblivingston committed:
revert vban

revert change from 0fc1994 'command, recorder, vban'

pblivingston committed:
Update higher.Tests.ps1

vban routes based on kind

pblivingston commented:
I've finished making changes here and all higher tests pass for all kinds. I expanded the tests for pretty exhaustive coverage, but I'll test again and manually test and document anything not covered by the pester tests with each incremental PR.

Do you have a preference for how the PR should be split? I know I'd want to add IRemote and ArrayMembers first, and if you don't have any preference, I'll just try to use my best judgement after that. Also, is the best way to do this to create branches based on 'dev' and simply copy changes from this branch?

I hope you can forgive me for stumbling around a bit as I learn!

onyx-and-iris commented:
Hi, alright thanks for letting me know. I was able to get basic and banana tests to pass but I had to increase the sleep time to 500ms on the following potato tests:

-   It 'Should set and get Option.monitoronsel'
-  It 'Should set and get Option.slidermode'
-  It 'Should set and get Recorder.channel'

---

IRemote would be a good place to start because it's entirely internal to the interface. Since ArrayMembers offers a clean decoupled interface for handling array like properties you can submit that on its own, otherwise submit it along with one of the new classes that implements it (Option or Patch?). I'd prefer to deal with changes that are additive and/or implement fixes that definitely don't require any breaking changes and to leave any changes that might require a breaking change until last. The two breaking changes I'm considering are: remove current {Strip}.gainlayer properties and remove {Recorder}.loop so perhaps we can leave Strip and Recorder classes until last.

Not required but feel free to raise an issue detailing the changes you intend to submit and then link a PR to it.

You could create several branches from dev but you might need to fast forward a branch if it falls behind dev (if some other branch gets merged into dev). Otherwise create branches from branches, for example, create a branch that adds-iremote, then from that create another branch that adds-arraymembers. Otherwise just wait for me to merge changes into dev and branch off that.


I hope that's helpful, thanks for your efforts.

pblivingston commented:
That's very helpful, thank you!

Would you prefer the higher.Tests.ps1 in each PR to only test changes made in that branch, or would you prefer it to be cumulative to ultimately provide an exhaustive test of the surface as in this branch? I could also do both and provide a changes.Tests.ps1 to reflect just the changes made in that branch.

Before adding IRemote, I want to make a handful of fixes to existing code that won't otherwise change - if you'd like cumulative tests, I can set the groundwork for that alongside the preliminary fixes (unless you think the test changes on their own would be significant enough for their own PR). I'll open an issue detailing the fixes with a note about tests.

onyx-and-iris commented:
The tests all pass in the dev branch in its current state. So what you'd do is create a PR with changes and add to the current tests (in higher.Tests.ps1). Diffs would look like this:

- new feature
- tests that test only the new feature

Do that for each new feature/addition.

If at the end you still want to add more tests please do so in its own PR. Strictly speaking test suites don't need to be exhaustive but I don't mind if you want to flesh them out.

pblivingston commented:
Okay, sounds good!
